src/Hl7.Fhir.Base/Model/Base.TypedElement.cs

```diff
@@ -0,0 +1,177 @@
+#if true
+
+#nullable enable
+
+using Hl7.Fhir.ElementModel;
+using Hl7.Fhir.Introspection;
+using Hl7.Fhir.Serialization;
+using Hl7.Fhir.Specification;
+using Hl7.Fhir.Utility;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using P=Hl7.Fhir.ElementModel.Types;
+
+namespace Hl7.Fhir.Model;
+
+
+/// <summary>
+/// An element within a tree of typed FHIR data with also a parent element.
+/// </summary>
+/// <remarks>
+/// This interface represents FHIR data as a tree of elements, including type information either present in
+/// the instance or derived from fully aware of the FHIR definitions and types
+/// </remarks>
+#pragma warning disable CS0618 // Type or member is obsolete
+public interface IScopedNode : ITypedElement, IShortPathGenerator
+#pragma warning restore CS0618 // Type or member is obsolete
+{
+    /// <summary>
+    /// The parent node of this node, or null if this is the root node.
+    /// </summary>
+    IScopedNode? Parent { get; } 
+
+    // /// <summary>
+    // /// An indication of the location of this node within the data represented by the <c>ITypedElement</c>.
+    // /// </summary>
+    // /// <remarks>The format of the location is the dotted name of the property, including indices to make
+    // /// sure repeated occurrences of an element can be distinguished. It needs to be sufficiently precise to aid
+    // /// the user in locating issues in the data.</remarks>
+    // string Location { get; }
+}
+
+internal record ScopeInformation(IScopedNode? Parent, string Name, int? Index);
+
+
+public abstract partial class Base : IScopedNode,
+    IFhirValueProvider, IResourceTypeSupplier
+{
+    [NonSerialized]
+    private ScopeInformation? _scopeInfo;
+    
+    private ScopeInformation ScopeInfo
+    {
+        get => LazyInitializer.EnsureInitialized(ref _scopeInfo, () => BuildRoot())!;
+        set => _scopeInfo = value;
+    } 
+
+    internal ScopeInformation BuildRoot(string? rootName = null) => new(null, rootName ?? TypeName, null);
+    
+    internal Base WithScopeInfo(ScopeInformation info)
+    {
+        this.ScopeInfo = info;
```

Comment:
We should create benchmarks to see if not having a record but separate private fields in Base is faster. This, after all, does create an object in the heap.

Reply:
I made it a record struct, that way it is on the stack :)

Reply:
How can it be? It's a member of an object (a POCO) that is on the heap?  I fear that there is a lot of copying going on then.

Reply:
Ah, we should not thing of structs as "stack types" but "value types".  The JIT compiler MAY put them on the stack, but doesn't always do so. In this case, it does not, and your ScopeInformation type gets value semantics. In other words on every function call with a parameter of type ScopeInformation, it is copied, on every assignment it is copied. And it ends up on the heap if it a property or member of a class. 

See https://learn.microsoft.com/en-us/archive/blogs/ericlippert/the-stack-is-an-implementation-detail-part-one.

So, definitely not make this a struct!

Reply:
It's still a struct....

Reply:
Oh well, ValueTuple is a struct too, I guess this is comparable.